server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java

```diff
@@ -65,47 +65,59 @@ public void writeTo(StreamOutput out) throws IOException {
 
     @Override
     public Diff<DiffableStringMap> diff(DiffableStringMap previousState) {
-        return new DiffableStringMapDiff(previousState, this);
+        if (this.equals(previousState)) {
+            return DiffableStringMapDiff.EMPTY;
+        }
+        final List<String> tempDeletes = new ArrayList<>();
+        final Map<String, String> tempUpserts = new HashMap<>();
+        for (String key : previousState.keySet()) {
+            if (containsKey(key) == false) {
+                tempDeletes.add(key);
+            }
+        }
+
+        for (Map.Entry<String, String> partIter : entrySet()) {
+            String beforePart = previousState.get(partIter.getKey());
+            if (beforePart == null) {
+                tempUpserts.put(partIter.getKey(), partIter.getValue());
+            } else if (partIter.getValue().equals(beforePart) == false) {
+                tempUpserts.put(partIter.getKey(), partIter.getValue());
+            }
+        }
+        return getDiff(tempDeletes, tempUpserts);
     }
 
     public static Diff<DiffableStringMap> readDiffFrom(StreamInput in) throws IOException {
-        return new DiffableStringMapDiff(in);
+        final List<String> deletes = in.readStringList();
+        final Map<String, String> upserts = in.readMap(StreamInput::readString, StreamInput::readString);
+        return getDiff(deletes, upserts);
+    }
+
+    private static Diff<DiffableStringMap> getDiff(List<String> deletes, Map<String, String> upserts) {
+        if (deletes.isEmpty() && upserts.isEmpty()) {
+            return DiffableStringMapDiff.EMPTY;
+        }
+        return new DiffableStringMapDiff(deletes, upserts);
     }
 
     /**
      * Represents differences between two DiffableStringMaps.
      */
     public static class DiffableStringMapDiff implements Diff<DiffableStringMap> {
 
-        public static final DiffableStringMapDiff EMPTY = new DiffableStringMapDiff(DiffableStringMap.EMPTY, DiffableStringMap.EMPTY);
+        public static final DiffableStringMapDiff EMPTY = new DiffableStringMapDiff(Collections.emptyList(), Collections.emptyMap()) {
```

Comment:
Nit: For consistency sake I think makes sense to use either (`List.of()` and `Map.of`) or  (`Collections.emptyList()` and `Collections.emptyMap()`) both in `DiffableUtils` and  `DiffableStringMap`.